aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -0,0 +1,154 @@
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.sync.ResponseTransformer;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+
+public class S3InputStream extends SeekableInputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3InputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final S3Client s3;
+  private final AmazonS3URI location;
+
+  private InputStream stream;
+  private long pos = 0;
+  private long next = 0;
+  private boolean closed = false;
+
+  private int skipSize = 1024 * 1024;
+
+  public S3InputStream(S3Client s3, AmazonS3URI location) {
+    this.s3 = s3;
+    this.location = location;
+
+    createStack = Thread.currentThread().getStackTrace();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "already closed");
+    Preconditions.checkArgument(newPos >= 0, "position is negative: %s", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    next = newPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    return stream.read();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "already closed");
+    positionStream();
+
+    int bytesRead = stream.read(b, off, len);
+    pos += bytesRead;
+    next += bytesRead;
+
+    return bytesRead;
+  }
+
+  @Override
+  public void close() throws IOException {
+    super.close();
+    closed = true;
+    closeStream();
+  }
+
+  private void positionStream() throws IOException {
+    if ((stream != null) && (next == pos)) {
+      // already at specified position
+      return;
+    }
+
+    if ((stream != null) && (next > pos)) {
+      // seeking forwards
+      long skip = next - pos;
+      if (skip <= Math.max(stream.available(), skipSize)) {
+        // already buffered or seek is small enough
+        LOG.debug("Read-through seek for {} to offset {}", location, next);
+        try {
+          if (stream.skip(skip) == skip) {
+            pos = next;
+            return;
+          }
+        } catch (IOException ignored) {
+          // will retry by re-opening the stream
+        }
+      }
+    }
+
+    // close the stream and open at desired position
+    LOG.debug("Seek with new stream for {} to offset {}", location, next);
+    pos = next;
+    openStream();
+  }
+
+  private void openStream() throws IOException {
+    GetObjectRequest request = GetObjectRequest.builder()
+        .bucket(location.getBucket())
+        .key(location.getKey())
+        .range(String.format("bytes=%s-", pos))
```

Comment:
This is weird. The format for this argument is `"bytes=(num)-"`?

Reply:
Yeah, I made a comment about this above as well. The v2 way they handle range via the API is based off of [rfc2616-sec14](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.35) and the API only exposes the ability to set this string.  It feels very awkward compared to the v1 API that exposes the range as `long` with options for start and end of range.  Weird, but I don't see a better option.